PULL_REQUEST_WORKFLOW.rst

```diff
@@ -0,0 +1,260 @@
+ .. Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+ ..   http://www.apache.org/licenses/LICENSE-2.0
+
+ .. Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+.. contents:: :local:
+
+Why non-standard pull request workflow
+--------------------------------------
+
+This document describes the Pull Request Workflow we've implemented in Airflow. The workflow is slightly
+more complex than regular workflow you might encounter in most of the projects because after experiencing
+some huge delays in processing queues in October 2020 with GitHub Actions, we've decided to optimize the
+workflow to minimize the use of Github Actions build time by utilising selective approach on which tests
+and checks in the CI system are run depending on analysis of which files changed in the incoming PR and
+allowing the Committers to control the scope of the tests during the approval/review process.
+
+Just to give a bit of context, we started off with the approach that we always run all tests for all the
+incoming PRs, however due to our matrix of tests growing, this approach did not scale with the increasing
+number of PRs and when we had to compete with other Apache Software Foundation projects for the 180
+slots that are available for the whole organization. More Apache Software Foundation projects started
+to use GitHub Actions and we've started to experience long queues when our jobs waited for free slots.
+
+We approached the problem by:
+
+1) Improving mechanism of canceling duplicate workflow runs more efficiently in case of queue conditions
+   (duplicate workflow runs are generated when someone pushes a fixup quickly - leading to running both
+   out-dated and current run to completion, taking precious slots. This has been implemented by improving
+   `cancel-workflow-run <https://github.com/potiuk/cancel-workflow-runs/>`_ action we are using. In version
+   4.1 it got a new feature of canceling all duplicates even if there is a long queue of builds.
```

Comment:
```suggestion
1) Improving mechanism of cancelling duplicate workflow runs more efficiently in case of queue conditions
   (duplicate workflow runs are generated when someone pushes a fixup quickly - leading to running both
   out-dated and current run to completion, taking precious slots. This has been implemented by improving
   `cancel-workflow-run <https://github.com/potiuk/cancel-workflow-runs/>`_ action we are using. In version
   4.1 it got a new feature of cancelling all duplicates even if there is a long queue of builds.
```